windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -0,0 +1,1348 @@
+    procedure Enter; override;
+    procedure Exit; override;
+    procedure HandleCheck; override;
+    function  HandleKmShell : Integer; override;
+    procedure HandleDownload; override;
+    procedure HandleAbort; override;
+    procedure HandleInstallNow; override;
+    function StateName: string; override;
+  end;
+
+  WaitingRestartState = class(TState)
+  public
+    procedure Enter; override;
+    procedure Exit; override;
+    procedure HandleCheck; override;
+    function  HandleKmShell : Integer; override;
+    procedure HandleDownload; override;
+    procedure HandleAbort; override;
+    procedure HandleInstallNow; override;
+    function StateName: string; override;
+  end;
+
+  InstallingState = class(TState)
+  private
+    procedure DoInstallKeyman; overload;
+    function DoInstallKeyman(SavePath: string) : Boolean; overload;
+    {
+      Installs the Keyman file using either msiexec.exe or the setup launched in
+      a separate shell.
+
+      @params  Package  The package to be installed.
+
+      @returns True  if the installation is successful, False otherwise.
+    }
+    function DoInstallPackage(Package: TUpdateStateMachineParamsPackage): Boolean;
+  public
+    procedure Enter; override;
+    procedure Exit; override;
+    procedure HandleCheck; override;
+    function  HandleKmShell : Integer; override;
+    procedure HandleDownload; override;
+    procedure HandleAbort; override;
+    procedure HandleInstallNow; override;
+    function StateName: string; override;
+  end;
+
+  RetryState = class(TState)
+  public
+    procedure Enter; override;
+    procedure Exit; override;
+    procedure HandleCheck; override;
+    function  HandleKmShell : Integer; override;
+    procedure HandleDownload; override;
+    procedure HandleAbort; override;
+    procedure HandleInstallNow; override;
+    function StateName: string; override;
+  end;
+
+  PostInstallState = class(TState)
+  private
+      procedure HandleMSIInstallComplete;
+  public
+    procedure Enter; override;
+    procedure Exit; override;
+    procedure HandleCheck; override;
+    function  HandleKmShell : Integer; override;
+    procedure HandleDownload; override;
+    procedure HandleAbort; override;
+    procedure HandleInstallNow; override;
+    function StateName: string; override;
+  end;
+
+
+  { This class also controls the state flow see  }
+  TUpdateStateMachine = class
+  private
+    FForce: Boolean;
+    FAutomaticUpdate: Boolean;
+    FParams: TUpdateStateMachineParams;
+    FErrorMessage: string;
+    DownloadTempPath: string;
+    FShowErrors: Boolean;
+    FDownload: TUpdateStateMachineDownloadParams;
+
+    CurrentState: TState;
+    // State object for performance (could lazy create?)
+    FIdle: IdleState;
+    FUpdateAvailable: UpdateAvailableState;
+    FDownloading: DownloadingState;
+    FWaitingRestart: WaitingRestartState;
+    FInstalling: InstallingState;
+    FRetry: RetryState;
+    FPostInstall: PostInstallState;
```

Comment:
These seem like they should be an array, e.g.:

```suggestion
  FStateInstance: TState[TUpdateState];
```

Then can use loops for initialization, destruction, etc., which is cleaner than referencing by name (and possibly missing one)